Framework/Algorithms/src/MultipleScatteringCorrection.cpp

```diff
@@ -0,0 +1,403 @@
+                  "Event workspace for multiple scattering correction.");
+}
+
+/**
+ * @brief validate the inputs
+ *
+ * @return std::map<std::string, std::string>
+ */
+std::map<std::string, std::string> MultipleScatteringCorrection::validateInputs() {
+  std::map<std::string, std::string> result;
+
+  // check 0: input workspace must have a valida sample
+  // NOTE: technically the workspace validator should be able to catch the undefined sample
+  //       error.  Keeping this check here in case the validator is changed in the future.
+  m_inputWS = getProperty("InputWorkspace");
+  const auto &sample = m_inputWS->sample();
+  if (!sample.getShape().hasValidShape()) {
+    result["InputWorkspace"] = "The input workspace must have a valid sample shape";
+  }
+
+  // check 1: input workspace must have a valid sample environment
+  // TODO: this check should be implemented once we start considering the multiple
+  // scattering correction between container and sample (heterogenous scattering media)
+  // NOTE: use PaalmanPingAbsorptionCorrection.cpp check as starting point
+
+  // others?
+
+  return result;
+}
+
+/**
+ * @brief execute the algorithm
+ *
+ */
+void MultipleScatteringCorrection::exec() {
+  // parse input properties and assign corresponding values to the member
+  // variables
+  parseInputs();
+
+  // prepare the cached distances
+  MultipleScatteringCorrectionDistGraber distGraber(m_inputWS->sample().getShape(), m_elementSize);
+  distGraber.cacheLS1(m_beamDirection);
+  // NOTE: the following data is now cached in the distGraber object
+  // std::vector<double> distGraber.m_LS1 : Cached L1 distances
+  // std::vector<double> distGraber.m_elementVolumes : Cached element volumes
+  // std::vector<Kernel::V3D> distGraber.m_elementPositions : Cached element positions
+  // size_t distGraber.m_numVolumeElements : The number of volume elements
+  size_t numVolumeElements = distGraber.m_numVolumeElements;
+
+  // L2D needs to be calculated w.r.t the detector
+  // L12 is independent from the detector, therefore can be cached outside
+  // - L12 is a upper off-diagonal matrix
+  // NOTE: if the sample size/volume is too large, we might need to use openMP
+  //       to parallelize the calculation
+  size_t len_l12 = numVolumeElements * (numVolumeElements - 1) / 2;
+  std::vector<double> sample_L12s(len_l12, 0.0);
+  for (size_t i = 0; i < numVolumeElements; ++i) {
+    for (size_t j = i + 1; j < numVolumeElements; ++j) {
+      const V3D dist = distGraber.m_elementPositions[i] - distGraber.m_elementPositions[j];
+      size_t idx = calcLinearIdxFromUpperTriangular(numVolumeElements, i, j);
+      sample_L12s[idx] = dist.norm();
+    }
+  }
+
+  // debug L12 matrix
+  // NOTE: this will come in handy when we start considering container and sample interaction,
+  //       do not remove it.
+  // std::ostringstream msg;
```

Comment:
It may be good to put this inside a `#ifndef NDEBUG` check so it is available in debug builds. Alternatively, changing the log level to `g_log.debug` would achieve the same thing, but may be slower.

Reply:
the macro approach looks more reasonable to me, implementing it now